linkcheck/url.py

```diff
@@ -91,6 +91,17 @@
 is_safe_fragment = re.compile("(?i)^%s$" % _safe_fragment_pattern).match
 
 
+def decode_for_unquote(part):
+    """
+    Decode string for unquote function
+    To string in Python 3, leave it in Python 2
+    """
+    try:  # Python 2
+        unicode
+        return part
+    except NameError:
+        return part.decode("utf-8")
```

Comment:
I'm not too happy about using NameError to distinguish Python versions.  How about

    if not isinstance(part, str):
        # Python 3: we probably got bytes
        part = part.decode("UTF-8")
    return part

?  Since we want `str` on both Python 2 and Python 3

Reply:
i'd also be worried about adding another function wrapper and exception handling in what is possibly a hot path. why are things in bytes in py3 and not py2? shouldn't this be decoded all the time?

Reply:
> why are things in bytes in py3 and not py2? shouldn't this be decoded all the time?

According to future/types/newbytes.py: "the Python 2 bytes object is a very, very
different beast to the Python 3 bytes object."

Reply:
okay then what's the way forward here?

Reply:
Option 1:
- the tests pass? ship it!

Option 2:
- somebody spends time coming up with alternative solutions

I don't have the energy for option 2 (sorry!), and while I'm not overjoyed with option 1, I will not object to it.

Python 2.7 will be EOLed Jan 1 2019.  We will serve our users better by having a version that works with Python 3.  If the new code turns out to cause problems not caught by tests, users will report them and we will fix them then (and, ideally, add tests).

Reply:
okay, so cjmayo - can you think of a nicer way of doing this? what do you think of @mgedmin's proposed solution?

i'm all for just plowing forward towards py3 as well, as stated elsewhere. :shipit:

Reply:
Option 1 I think. I'll put more of a comment on #228.

Reply:
> ```
> if not isinstance(part, str):
>     # Python 3: we probably got bytes
>     part = part.decode("UTF-8")
> return part
> ```

Testing with Python 2.7 and 3.6 lead me to:
```py
if not isinstance(part, str) and not isinstance(part, str_text):
```
Not pretty but removable along with Python 2 support.

Reply:
Seems to be causing TestHttpMisc.swf_test to fail.

Simplified:
```
>>> a = b"q\xfc'4309 uq0'34 tuq\xfct9 qe\xfcrgjsad\xf6fg jias#http://www.example.org/\xb0\xb0\n"
>>> a.decode("utf8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xfc in position 1: invalid start byte
```

Reply:
I've updated to:
```py
part = part.decode("UTF-8", errors="replace")
```
and all the tests are passing. Not great either, but my only idea is to use requests to do the downloading in UrlBase and derivatives.

Reply:
> if not isinstance(part, str) and not isinstance(part, str_text):

I believe this can be simplified to

```python
if not isinstance(part, (str, str_text)):
```

Reply:
That's neater. Implemented.